### PR TITLE
Sessions this CLI writes stay visible to the 3.x CLI sharing the store

### LINKED ISSUE #204: Credential store incompatibility: sessions refreshed by the rc CLI are invisible to @prisma/cli@latest (3.0.0-beta.30)
## Summary

The rc-line CLI and the 3.x beta CLI share the same auth store (`~/Library/Application Support/prisma/auth.json` + `auth.context.json` on macOS), but after the rc CLI refreshes a session, the 3.x CLI reports it as unauthenticated. Anyone using both CLIs on one machine (which the docs currently require: the homepage quickstarts use `@latest app deploy`, the Compute docs use `@next`) loses 3.x auth.

## Repro (macOS, Node 24.11.1)

1. Have a stored OAuth session for the active workspace (originally created via `auth login`).
2. Run any rc CLI command that refreshes the token, e.g. `prisma auth whoami` (8.0.0-rc.4, repo-built since published rc crashes — see #202). It reports `authenticated: true, source: "stored"` and rewrites the session in `auth.json`.
3. Immediately run `npx @prisma/cli@latest auth whoami --json` (3.0.0-beta.30):

```json
{ "ok": true, "result": { "authenticated": false, "provider": null, "user": null, "workspace": null, "credential": null } }
```

4. `npx @prisma/cli@latest app deploy … --json --no-interactive` then fails `AUTH_REQUIRED`, even though the rc CLI on the same machine is authenticated and the session entry for the active workspace exists in `auth.json` (verified: 5 sessions in the file, incl. the active workspace with a fresh `expiresAt`).

3.x's `FileTokenStorage.getTokens()` swallows all errors (`dist/adapters/token-storage.js`, catch → null), so whatever it rejects about the rc-written record is silent. Workaround that unblocked validation: `PRISMA_SERVICE_TOKEN`, which 3.x honors.

## Expected

Either both CLIs read/write a mutually compatible credential record, or the incompatibility is detected and surfaced ("session was written by a newer CLI, run auth login") instead of a silent `authenticated: false`.

Found while validating the docs against both CLI lines (prisma/web#8151).

🤖 Filed with [Claude Code](https://claude.com/claude-code) on behalf of datta@prisma.io

### PR DESCRIPTION
**Pull request by ankur-arch**
## What this PR does

Sessions written by this CLI stay visible to the 3.x CLI that shares the same auth store. Today our first write silently logs the 3.x CLI out. Merging this closes #204.

## The bug

Both CLI lines point at one auth store (`auth.json` plus `auth.context.json`) but disagree about its shape:

- The 3.x CLI (`@prisma/cli@latest`, 3.0.0-beta.30) reads sessions from a top-level `tokens` array (`data.tokens || []`), selected by the context file's `activeWorkspaceId`.
- This CLI writes `{ version, sessions, currentWorkspaceId }`, and it writes the whole file.

Adoption of the legacy store is deliberately a pure read, so 3.x sessions keep working until this CLI's first mutation. For someone who only runs read commands, that first mutation is the background token refresh. After it, 3.x reports `authenticated: false` with no error anywhere: its reader simply finds an empty array.

## The fix

One write choke point gains a legacy mirror. `writeCredentialState`:

1. also serializes the sessions in the legacy record shape under `tokens` (`{ workspaceId, token, refreshToken? }`), and
2. keeps `auth.context.json`'s `activeWorkspaceId` in step with `currentWorkspaceId`, preserving the remembered-workspace name map.

Every mutation flows through this function (login, refresh, select, end session, logout), so all of them stay legacy-visible. The mirror is invisible to this CLI's own reader, which branches on `sessions` before ever looking at `tokens`. This is also the behavior the code already intended: `auth/operations.ts` carries an unwired `storeLegacyCredential` helper written for exactly this purpose.

Hardening from the review rounds: the context file writes via temp plus rename (a torn context makes 3.x silently self-activate its latest session), and an empty state with no pre-existing context file does not materialize one (an existing null pointer reads as "signed out" to 3.x, an absent file does not).

## Proof

End-to-end against the real published 3.x binary, using a store refreshed through `activeCredentialStorage().setTokens` (the exact write the background refresh performs):

```
store written by main:        auth whoami -> authenticated: false
store written by this branch: auth whoami -> authenticated: true, workspace wksp_e2e
```

Regression tests read the store exactly as the 3.x CLI does: refresh keeps the session visible, create and select move the pointer, endSession preserves or clears the legacy view, and the mirror stays invisible to our own reader. The core two fail without the fix. Full suite: 985 passing, `tsc --noEmit` and `pnpm lint` clean.

## Scope

Two small functions in `legacy-state.ts` (the module that owns legacy-format knowledge) plus two lines in `writeCredentialState`. No changes to the credential manager, the lock protocol, or the schema this CLI reads. Known pre-existing gaps (the `project transfer` path writing through `@prisma/credentials-store`, dead `performLogout` code) are documented in the PR comments as follow-ups.

Fixes #204.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by ankur-arch** [comment]
End-to-end validation against the real published 3.x CLI, beyond the unit-level reader simulation:

Procedure: create a session and then rotate it through `activeCredentialStorage().setTokens` (the exact write the background refresh performs) into a temp store via `PRISMA_AUTH_FILE`, then run the packed `@prisma/cli@3.0.0-beta.30` binary against that store via `PRISMA_COMPUTE_AUTH_FILE`.

```
== store written by main's code ==
auth whoami --json → authenticated: false | workspace: null    (the #204 symptom)

== store written by this branch ==
auth whoami --json → authenticated: true  | workspace: { id: "wksp_e2e", ... }
```

**Comment by ankur-arch** [comment]
A fresh-context review of this diff surfaced two hardening items, now pushed: the auth.context.json write goes through temp+rename (a torn context makes the 3.x CLI silently self-activate its latest session), and an empty state with no pre-existing context file no longer materializes one (an existing null pointer reads as 'explicitly signed out' to 3.x, where an absent file lets it self-activate — an rc store that never had a context shouldn't flip that). Plus two more regression tests (pointer move preserves the workspace name map; empty write creates no context file). 983 tests pass.

Noted for follow-up, deliberately out of this PR's scope (pre-existing): the `project transfer` recipient path still writes the store through `@prisma/credentials-store` (legacy envelope, non-atomic, outside the state lock) — after it runs, rc falls back to legacy adoption; and `performLogout`/`storeLegacyCredential`/`guard.ts` are dead code. Both predate this change.

File: packages/cli/src/auth/legacy-state.ts [hunk 1]
```diff
@@ -5,6 +5,65 @@ import { getAuthContextFilePath } from "./token-storage";
...
+ * preserved verbatim; only the pointer moves.
+ */
+export async function syncLegacyContext(
+  authFilePath: string,
+  currentWorkspaceId: string | null,
+): Promise<void> {
+  const contextFilePath = getAuthContextFilePath(authFilePath);
+  const context = await readLegacyContext(contextFilePath);
+  if (context.exists && context.activeWorkspaceId === currentWorkspaceId) {
+    return;
```
[the remark this answers is not in the document]
Comment (ankur-arch) [reply packages/cli/src/auth/legacy-state.ts:43]: Verified against the actual 3.x reader before deciding: its `readAuthContext` **normalizes the pointer the same way** (`parsed.activeWorkspaceId.trim()` — `dist/adapters/token-storage.js`, same normalization this repo's `readLegacyContext` ports), so a whitespace-padded pointer is not invisible to it and the early-return cannot strand 3.x. The premise "the legacy reader uses the raw pointer" doesn't hold for the shipping 3.x build. The secondary point (malformed `workspaces` data isn't repaired while the pointer matches) is real but cosmetic — the names map is display-only and both readers tolerate it — so the guard stays: it keeps read-mostly mutations from rewriting a file they didn't change. Skipping this one.

File: packages/cli/tests/credential-manager-migration.test.ts [hunk 2]
```diff
@@ -291,3 +291,105 @@ describe("adopting the legacy store", () => {
...
+
+    expect((await readAsLegacyCli())?.workspaceId).toBe(WORKSPACE_B);
+
+    await manager.selectSession(WORKSPACE_A);
+    expect(await readAsLegacyCli()).toEqual({
+      workspaceId: WORKSPACE_A,
+      accessToken: tokenA,
+      refreshToken: "ra",
+    });
+  });
```
[the remark this answers is not in the document]
Comment (ankur-arch) [reply packages/cli/tests/credential-manager-migration.test.ts:378]: Added in the latest push: ending a non-active session keeps the remaining session and pointer visible to the 3.x reader, and ending the active session empties the mirror and nulls the pointer. 985 tests pass.

**Review by wmadden — APPROVED** [review]
